[ActionList] [Item] Deprecate ellipsis and introduce a truncate prop

zakwarsame opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [ComponentName] (if applicable), for example: [Button]
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Prefix it with [WIP] while it’s a work in progress
-->

### WHY are these changes introduced?

Resolves https://github.com/Shopify/core-workflows/issues/881
Fixes https://github.com/Shopify/core-workflows/issues/930

The `ellipsis` prop currently doesn't do anything more than adding `...` to the content string. This functionality doesn't achieve much when the intention is to actually `truncate` the content string

### WHAT is this pull request doing?

<!--
  Summary of the changes committed.

  Before / after screenshots are appreciated for UI changes. Make sure to include alt text that describes the screenshot.

  If you include an animated gif showing your change, wrapping it in a details tag is recommended. Gifs usually autoplay, which can cause accessibility issues for people reviewing your PR:

    <details>
      <summary>Summary of your gif(s)</summary>
      <img src="..." alt="Description of what the gif shows">
    </details>
-->

This PR introduces `truncate` as part of the `ActionListDescriptors` props. We can choose between `end` and `middle` truncation which is then dynamically applied to the item content. The function `middleTruncate` shortens the string based on character count.

We also set the `TopBar` user menu to a fixed width based on the mockups available on the original issue.

<!-- ℹ️ Delete the following for small / trivial changes -->

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

- [Spin URL](https://admin.web.actionlist.zakaria-warsame.us.spin.dev/store/shop1)
  - Notice no changes made to the top bar user menu first
  - Enable `new_account_menu_and_store_switcher` beta and refresh the page
  - Click on the top bar user menu
  - The third store has `middle` truncate applied and the fourth `end`
  - Notice the width is now `360` px

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

zakwarsame committed:
depracate ellipsis and introduce a truncate prop

zakwarsame committed:
changeset and update truncation implementation

update changeset

set default truncate to undefined

add test to check when truncate is absent

import groups order

return fullheight

introduce hasCustomActivator boolean to check for old activator

introduce hasCustomActivator boolean to check for old activator

update truncation implementation

changeset

update truncation implementation

update changeset

fixup

zakwarsame commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -134,3 +142,58 @@ export function Item({
   );
 }
+
+export const TruncateText = ({
```
This component could be top-level or maybe replace [Truncate](https://github.com/Shopify/polaris/tree/main/polaris-react/src/components/Truncate)

zakwarsame commented on polaris-react/src/components/TopBar/components/Menu/Menu.tsx:
```diff
@@ -77,8 +83,10 @@ export function Menu(props: MenuProps) {
     >
-      <ActionList onActionAnyItem={onClose} sections={actions} />
-      {messageMarkup}
+      <Box width={hasCustomActivator ? undefined : USER_MENU_POPOVER_WIDTH}>
```
This is based on [the new designs](https://www.figma.com/file/M2BnURpiYPiTT9z1D2iFNi/Enterprise-(CCS)-Nav?node-id=6147-204586&t=2iD2UT8l9BT6ip72-0). `hasCustomActivator` helps determine if we're passing the old user menu or the new one, and we adjust the width based on that.

Some stores will still keep the old user menu which is why we're doing the check.

zakwarsame committed:
update middle truncate because tooltip only allows for span

zakwarsame commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -89,7 +97,7 @@ export function Item({
 
   const contentElement = (
-    <HorizontalStack blockAlign="center" gap="4">
+    <HorizontalStack blockAlign="center" gap="4" wrap={!truncate}>
```
There could be use cases outside truncation where we want `ActionList` to not wrap. @emma-boardman's [PR](https://github.com/Shopify/polaris/pull/9029/files) proposes a `prop` to customize this. We'll still want wrap to be `false` for truncation on top of that prop.

kyledurand commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -134,3 +142,58 @@ export function Item({
+
+  return isOverflowing ? (
+    <Tooltip
+      zIndexOverride={514}
```
You should pull this value from tokens: `tokens.zIndex['z-6']`

kyledurand commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -134,3 +142,58 @@ export function Item({
+    </span>
+  );
+
+  const truncatedText = middleTruncate(children, 40);
```
Can we do something like:

`const MIDDLE_TRUNCATION_LENGTH = 40;`

At the top of the file?

kyledurand reviewed (changes requested):
I'm not a huge fan of needing these expensive measurements for middle truncation over just using css. 

Middle truncation is something we've decided against in the past for performance and accessibility reasons. 

What's the rationale for middle truncation and can we push back on it to just use css?

zakwarsame committed:
review: use only css for truncation

zakwarsame committed:
review: use only css for truncation

zakwarsame committed:
review: use only css for truncation

zakwarsame committed:
review: remove unused vals

kyledurand commented on polaris-react/src/components/ActionList/components/Item/tests/Item.test.tsx:
```diff
@@ -102,6 +102,22 @@ describe('<Item />', () => {
+    const item = mountWithApp(<Item content="Test" truncate={false} />);
+    expect(item).not.toContainReactComponent(Text, {
+      children: 'Test',
+      truncate: true,
```
```suggestion
    expect(item).toContainReactComponent(Text, {
      children: 'Test',
      truncate: false,
```

kyledurand commented on polaris-react/src/components/TopBar/components/Menu/Menu.tsx:
```diff
@@ -25,6 +28,8 @@ export interface MenuProps {
   /** A string that provides the accessibility labeling */
   accessibilityLabel?: string;
+  /** A boolean property indicating whether a custom activator is being used */
+  hasCustomActivator?: boolean;
```
What if this was `customActivatorWidth` so we could remove the magic number in `USER_MENU_POPOVER_WIDTH`?

zakwarsame replied:
I think that makes more sense than a fixed width, especially since the design was updated to have a `300px` instead. We could also make it more general to `customWidth` since the internal activator is now default to all stores.

kyledurand reviewed (approved):
This looks great, just a couple of minor comments. Not sure if my suggestion would work or not but if so it would allow for more flexibility

zakwarsame committed:
review: update stories to open by default; update customWidth; update menu item heights;

remove item from story and update comments

move store stories to WithMessage

update accessibility

update class name

henryyi commented on .changeset/dull-pumas-wave.md:
```diff
@@ -0,0 +1,6 @@
+'@shopify/polaris': minor
+---
+
+- Deprated ellipsis and introduced a truncate prop to `ActionList` items.
```
there's a typo: `Deprecated`

henryyi commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -61,7 +62,16 @@ export function Item({
+  let contentText: string | React.ReactNode = content || '';
+  if (ellipsis) {
+    contentText = `${content}…`;
+  } else if (truncate && content) {
```
nit: since we're deprecating `ellipsis` I think `truncate` should take precedence over it.

rcaplanshopify reviewed (approved):
Code and 🎩 LGTM!

henryyi commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -61,7 +62,16 @@ export function Item({
+  } else if (truncate && content) {
+    contentText = (
+      <Text as="span" variant="bodyMd" truncate>
+        {content}
```
Do we need a tooltip when it truncates?

zakwarsame committed:
review: type and if order

zakwarsame committed:
return tooltip but only when the truncate prop is passed

alex-page commented on polaris-react/src/components/ActionList/ActionList.scss:
```diff
@@ -129,3 +129,10 @@
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+}
```
I would prefer to use the `<Text>` component and not add duplicate styles here.

zakwarsame committed:
use Text component for truncation instead of cusotm styles

zakwarsame committed:
make span to display block

zakwarsame committed:
remove duplicates from text component

zakwarsame committed:
use Box instead of regular span

zakwarsame committed:
implement zIndex using polaris tokens
